### PR TITLE
chore: refactor ResolveAutostart tests to use dbfake

### PR DESCRIPTION
**Pull request by sreya**

### CODE REVIEW

File: coderd/database/dbfake/dbfake.go [hunk 1]
```diff
@@ -37,6 +37,64 @@ func Workspace(t testing.TB, db database.Store, seed database.Workspace) databas
...
+		Time:  dbtime.Now(),
+		Valid: true,
+	})
+
+	job = dbgen.ProvisionerJob(t, db, nil, job)
+	ProvisionerJobResources(t, db, job.ID, "", resources...)
+	return templateVersion
+}
+
+func TemplateVersionWithParams(t testing.TB, db database.Store, tv database.TemplateVersion, job database.ProvisionerJob, params []database.TemplateVersionParameter) (database.TemplateVersion, []database.TemplateVersionParameter) {
```
Comment (spikecurtis) [inline coderd/database/dbfake/dbfake.go:81]: You've got `TemplateVersion` (which accepts optional resources) and `TemplateVersionWithParams` (which accepts params, but no resources).  And, most of the time you pass an empty `database.TemplateVersion` and `database.ProvisionerJob`.  What do you think of making the version, job, resources and parameters optional parameters, either with `...opts` or using a Builder pattern?

File: coderd/database/dbfake/dbfake.go [hunk 2]
```diff
@@ -56,6 +114,19 @@ func WorkspaceWithAgent(t testing.TB, db database.Store, seed database.Workspace
 	return ws, authToken
 }
 
+func WorkspaceBuildWithParameters(t testing.TB, db database.Store, ws database.Workspace, build database.WorkspaceBuild, params []database.WorkspaceBuildParameter, resources ...*sdkproto.Resource) (database.WorkspaceBuild, []database.WorkspaceBuildParameter) {
```
Comment (spikecurtis) [inline coderd/database/dbfake/dbfake.go:117]: same idea here: make resources and parameters options of `WorkspaceBuild()`

File: coderd/database/dbgen/dbgen.go [hunk 3]
```diff
@@ -229,28 +229,63 @@ func WorkspaceBuild(t testing.TB, db database.Store, orig database.WorkspaceBuil
...
+		})
+		if err != nil {
+			return err
+		}
+
+		params, err = tx.GetWorkspaceBuildParameters(Ctx, id)
+		if err != nil {
+			return err
+		}
+		return err
```
Comment (spikecurtis) [inline coderd/database/dbgen/dbgen.go:261]: `if` test above is redundant

File: enterprise/coderd/workspaces_test.go [hunk 4]
```diff
@@ -1074,7 +1075,7 @@ func TestWorkspaceLock(t *testing.T) {
 func TestResolveAutostart(t *testing.T) {
 	t.Parallel()
 
-	ownerClient, owner := coderdenttest.New(t, &coderdenttest.Options{
+	ownerClient, db, owner := coderdenttest.NewWithDatabase(t, &coderdenttest.Options{
 		Options: &coderdtest.Options{
 			IncludeProvisionerDaemon: true,
```
Comment (spikecurtis) [inline enterprise/coderd/workspaces_test.go:1080]: remove provisioner daemon

File: coderd/database/dbfake/dbfake.go [hunk 5]
```diff
@@ -161,42 +164,18 @@ func (b WorkspaceBuildBuilder) Do() database.WorkspaceBuild {
...
-		err = b.ps.Publish(codersdk.WorkspaceNotifyChannel(build.WorkspaceID), []byte{})
+		err = b.ps.Publish(codersdk.WorkspaceNotifyChannel(resp.Build.WorkspaceID), []byte{})
 		require.NoError(b.t, err)
 	}
-	return build
+
+	for i := range b.params {
+		b.params[i].WorkspaceBuildID = resp.Build.ID
+	}
+	_ = dbgen.WorkspaceBuildParameters(b.t, b.db, b.params)
```
Comment (spikecurtis) [inline coderd/database/dbfake/dbfake.go:177]: I think this should happen before `Publish()` to the pubsub, to more closely match what provionerd would do; e.g. some subscribers might expect the parameters to have been set, and would race against your insert here.

**Review by spikecurtis — APPROVED** [review]
LGTM, suggestions inline, but I don't need to review again

File: coderd/database/dbfake/dbfake.go [hunk 6]
```diff
@@ -16,37 +16,65 @@ import (
...
 	"github.com/coder/coder/v2/coderd/database/pubsub"
 	"github.com/coder/coder/v2/coderd/provisionerdserver"
+	"github.com/coder/coder/v2/coderd/rbac"
 	"github.com/coder/coder/v2/coderd/telemetry"
 	"github.com/coder/coder/v2/codersdk"
 	sdkproto "github.com/coder/coder/v2/provisionersdk/proto"
 )
 
-type WorkspaceBuilder struct {
+var sysCtx = dbauthz.As(context.Background(), rbac.Subject{
```
[the remark this answers is not in the document]
Comment (sreya) [reply coderd/database/dbfake/dbfake.go:25]: Trying to run `db.UpdateActiveTemplateVersion` doesn't work with a system ctx.

**Pull request merged**